Improved back pressure

pveentjer opened the pull request.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java:
```diff
@@ -216,6 +210,23 @@
+     * number of concurrent invocations has exceeded and a new invocation comes in, then an exponential back-off is applied
+     * till eventually a timeout happens or there is room for the invocation.
+     *
+     * By default is is configured as 100, so with 271 partitions that would give (271+1)*100=27200 concurrent invocations from a
```
"By default is is" vs. "it is"

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/BackpressureRegulator.java:
```diff
@@ -0,0 +1,200 @@
+
+
+/**
+ * The InvocationBackPressureRegulator is responsible for regulating invocation 'pressure' and if sees that the system
```
Javadoc needs some fixes. Two newlines before Javadoc.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/BackpressureRegulator.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    /**
+     * Checks if a sync is forced for
```
Unfinished sentence, incomplete JavaDoc (parameter, return).

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java:
```diff
@@ -0,0 +1,246 @@
+
+        public CallIdSequenceWithBackpressure(int maxConcurrentInvocations) {
+            if (maxConcurrentInvocations < 1) {
+                throw new IllegalArgumentException("maxConcurrentInvocations can't be smaller than 0");
```
Code checks for "value < 1", exception message says "value smaller than 0"

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java:
```diff
@@ -0,0 +1,246 @@
+import static com.hazelcast.spi.OperationAccessor.setCallTimeout;
+
+/**
+ * Responsible for generating call id's.
```
No apostrophe for plural.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java:
```diff
@@ -0,0 +1,246 @@
+ * Responsible for generating call id's.
+ * <p/>
+ * It is very important that for each {@link #next(Invocation)}, there is a matching {@link #complete(Invocation)}.
+ * If this doesn't match, than either the number of concurrent invocations can grow or shrink ever time. This can lead
```
ever/over time

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java:
```diff
@@ -0,0 +1,246 @@
+ * <p/>
+ * In the future we could add a CallIdSequence per partition or using some 'concurrency level' and do a mod based on the
+ * partition-id. The advantage is that you reduce contention and improved isolation, at the expensive of increased complexity
+ * + not always being able to fully utilize the number of invocations.
```
I think this can be deleted.

pveentjer replied:
which part?

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java:
```diff
@@ -0,0 +1,246 @@
+     */
+    public static final class CallIdSequenceWithBackpressure extends CallIdSequence {
+        private static final int MAX_DELAY_MS = 500;
+        private static final int INDEX_HEAD = 7;
```
What's the reason for beginning with 7?

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java:
```diff
@@ -126,39 +127,52 @@
+            /*
+             * final long minTimeout = Math.min(defaultCallTimeout, MIN_TIMEOUT);
+             * long callTimeout = Math.min(waitTimeoutMillis, defaultCallTimeout);
+             * callTimeout = Math.max(a, minTimeout);
```
I think `a` should be `callTimeout` in your example.

pveentjer replied:
I only cleaned up the structure of the code. So the wait timeout was already there. It isn't changed by me.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationRegistry.java:
```diff
@@ -40,99 +37,80 @@
+     * If the invocation registration was skipped, the call is ignored.
      *
      * @param invocation the Invocation to deregister.
      */
```
`@throws AssertionError if deregister of invocation failed` ?

pveentjer replied:
I'm not that crazy about documenting asserts.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationRegistry.java:
```diff
@@ -356,10 +299,38 @@ private void scanHandleOperationTimeout() {
                 try {
                     invocation.checkBackupTimeout(backupTimeoutMillis);
                 } catch (Throwable t) {
+                    t.printStackTrace();
```
Superfluous `t.printStackTrace()`? Should be printed via `logger.severe()`.

pveentjer replied:
Good one.. I'll fix that.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java:
```diff
@@ -295,20 +320,20 @@ public void logOperationDeserializationException(Throwable t, long callId) {
-            final Level level = returnsResponse ? Level.FINEST : Level.WARNING;
-            if (operationService.logger.isLoggable(level)) {
-                operationService.logger.log(level, t.getClass().getName() + ": " + t.getMessage());
+            final Level level = returnsResponse ? FINEST : Level.WARNING;
```
static import for `Level.WARNING`?

pveentjer replied:
static imports.

mdogan commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationRegistry.java:
```diff
@@ -40,99 +37,80 @@
         }
 
-        invocations.remove(callId);
+        assert invocations.remove(callId) != null : "failed to deregister callId:" + callId + " " + invocation;
```
Are you sure `invocations.remove(callId)` will be called when assertions are disabled (which is default)? AFAIK it will not.

mdogan replied:
try this:

```java
static boolean FOO_CALLED;

    public static void main(String[] args) throws Exception {
        assert foo();
        System.out.println(FOO_CALLED);

    }

    static boolean foo() {
        FOO_CALLED = true;
        return true;
    }
```

pveentjer replied:
That is a good one! 

Initially I didn't have the assert.. but I ran into some test failures and added the assert around it.  So I'll fix that asap.

mdogan commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/BackpressureRegulator.java:
```diff
@@ -0,0 +1,204 @@
+
+/**
+ * The BackpressureRegulator is responsible for regulating invocation 'pressure' and if sees that the system
+ * is getting overloaded, it will apply back pressure, so the the system
```
It looks like sentence 
> ... so the the system

is not complete here...

mdogan commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/BackpressureRegulator.java:
```diff
@@ -0,0 +1,204 @@
+ * The BackpressureRegulator is responsible for regulating invocation 'pressure' and if sees that the system
+ * is getting overloaded, it will apply back pressure, so the the system
+ * <p/>
+ * Responsible for regulating invocation pressure on the Hazelcast system.cIn most cases the invocations on Hazelcast
```
A typo
> on the Hazelcast system.cIn most cases

*on the Hazelcast system. In most cases*

mdogan commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java:
```diff
@@ -0,0 +1,249 @@
+
+    public static final class CallIdSequenceWithoutBackpressure extends CallIdSequence {
+
+        private static final AtomicLongFieldUpdater<CallIdSequenceWithoutBackpressure> HEAD
```
What is the advantage of using `AtomicLongFieldUpdater` here? Why not regular `AtomicLong`?

pveentjer replied:
One less memory indirection.

pveentjer replied:
Strange.. I thought I left a comment here.

There is one less memory indirection. I don't want to make the previous situation slower than the current one. So instead of having an atomiclong with a volatile field, you have a CallSequenceIdWithoutBackPressure with volatile field.

mdogan commented on hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImplTest.java:
```diff
@@ -0,0 +1,300 @@
+        verify(responseHandler).sendResponse(any(ExpectedRuntimeException.class));
+    }
+
+//    @Test
```
Are these tests commented out intentionally or are they left-over?

pveentjer replied:
I'll check them.

mdogan commented on hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/OperationServiceImpl_invokeOnTargetTest.java:
```diff
@@ -81,6 +81,7 @@ public void test_nonExistingTarget() throws UnknownHostException {
         }
     }
 
+    //todo
```
What about adding `@Ignored("TODO: will implement later...")` to make it more visible? Currently this's an empty method.

pveentjer replied:
Good one.

mdogan commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java:
```diff
@@ -0,0 +1,249 @@
+                    remainingTimeoutMs -= delayMs;
+                }
+            } finally {
+                setCallTimeout(invocation.op, remainingTimeoutMs);
```
I'm not sure if setting remaining time as call-timeout. Because call-timeout is used as timeout of each remote call, it means for each retry not the whole invocation. 

For example normally, using 1min call-timeout, a caller can wait infinitely as long as a retry-response is received in every 1min. But when no response is received in call-timeout (+ some extra time), invocation will throw operation-timeout exception. 

If you set remaining-time as call-timeout, on every retry call-timeout will decrease until it reaches zero. I don't know if this is your intention...

pveentjer replied:
It was my intention. But I know this issue with blocking calls because they interact in a complex way with the timeout. 

What is your suggestion? 

I can introduce a new timeout: wait for call invocation slot timeout 

It will be be deducted from your call timeout. It is just some global constant (through properties) that is the same for every invocation.

mdogan replied:
A new timeout will make it more complex, I think. It's already complex, I've never managed to make it simpler since Hazelcast 1. 

What about storing original call-timeout in invocation and using it for slot timeout and restoring operation call-timeout on every retry?

pveentjer replied:
This new timeout will not be registered on the operation or invocation btw. It is just some kind of timeout used somewhere in the system. The only problem is that a user has no control on this overload timeout apart from a groupsproperty. The time spend on doing waiting will also not be deducted anywhere.. 

That is what I meant with an additional property.

mdogan replied:
ok, fine for me.

pveentjer replied:
Thanks. I'll do  that. We can always come up with a different mechanism later.

pveentjer commented:
I have updated a PR. I have added a backoff timeut configurable through the GroupsProperties. No more messing around with the timeout of the Operation.

mdogan commented:
:+1:

pveentjer commented:
@gurbuzali @mdogan how to proceed? Someone is going to hit the merge button?

gurbuzali commented:
I'll hit the green button in a minute, reviewing myself already

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/BackpressureRegulator.java:
```diff
@@ -0,0 +1,214 @@
+
+        AtomicInteger syncDelayCounter = syncDelayCounter(op);
+
+        for (; ; ) {
```
If I'm not totally wrong this loop is not wait-free, isn't it? It does not return after a defined number of tries, theoretically it could loop for ever.

A lot of loops would just happen under extreme load. In this case I would not worry that there is no forced sync operation being created. So a solution for this might be to return with `syncForced = false` if you looped for more tries than the first `syncDelayCounter.get()` was. After so many tries another backup operation did the sync and we could happily retreat with async (and even without decreasing the counter).

Donnerbart replied:
Or maybe better: check if the `prev` is higher than in the last iteration. Then we can be sure another operation was forced to be sync.

pveentjer replied:
I'm not worried about this not being wait free because:
- a thread needs to do so much other stuff to complete a backup, that there is a very small chance that 2 threads at the same moment are going to content with each other
- there are 271 partitions by default and each partition will get its own sync delay counter. This means that there is a very small chance that 2 threads are going to content for the same counter.

Donnerbart replied:
Fair enough. I'm just a bit distrustful whenever I see a `for (; ; )` or `while (true)` :)

pveentjer replied:
That is a typical way to set up a non blocking algorithm. I'm not distrustful about it.
